### PR TITLE
Backend auth app

### PR DESCRIPTION
**Pull request by Eugineer0**
Minimal authentication backend-app with access JWT and DB user storage

### CODE REVIEW

File: Backend/CarRentalApp/CarRentalApp/Program.cs [hunk 1]
```diff
@@ -1,23 +1,65 @@
...
+
+var accessJwtValidationParams = accessJwtConfig.ValidationParameters;
+accessJwtValidationParams.IssuerSigningKey = new SymmetricSecurityKey(
+    Encoding.UTF8.GetBytes(accessJwtConfig.GenerationParameters.Secret)
+);
+
+builder.Services.AddAuthentication(JwtBearerDefaults.AuthenticationScheme).AddJwtBearer(options =>
+{
+    options.TokenValidationParameters = accessJwtValidationParams;
+});
```
Comment (andrei-kruhliakou-itechart) [inline Backend/CarRentalApp/CarRentalApp/Program.cs]:
```suggestion
builder.Services
    .AddAuthentication(JwtBearerDefaults.AuthenticationScheme)
    .AddJwtBearer(options =>
        {
            options.TokenValidationParameters = accessJwtValidationParams;
        }
    );
```

File: Backend/CarRentalApp/CarRentalApp/Program.cs [hunk 2]
```diff
@@ -1,23 +1,65 @@
...
 {
-    
-}
+    options.AddDefaultPolicy(
+        builder =>
+        {
+            builder.AllowAnyOrigin()
+                .AllowAnyMethod()
+                .AllowAnyHeader();
+        });
```
Comment (andrei-kruhliakou-itechart) [inline Backend/CarRentalApp/CarRentalApp/Program.cs]:
```suggestion
    options.AddDefaultPolicy(
        builder =>
        {
            builder
                .AllowAnyOrigin()
                .AllowAnyMethod()
                .AllowAnyHeader();
        }
    );
```

Comment (andrei-kruhliakou-itechart) [reply Backend/CarRentalApp/CarRentalApp/Program.cs]: Suggested code style:
- place ALL chained method calls on the same level;
- place closing bracket on the same level with the statement that contains opening bracket;

File: Backend/CarRentalApp/CarRentalApp/Services/Identity/UserService.cs [hunk 3]
```diff
@@ -0,0 +1,33 @@
...
+            _userRepository = userRepository;
+            _passwordService = passwordService;
+        }
+
+        public async Task<User?> GetExistingUserAsync(UserLoginDTO user2Login)
+        {
+            return await _userRepository.GetByUsernameAsync(user2Login.Username);
+        }
+
+        public bool Validate(User existingUser, UserLoginDTO user2Validate)
```
Comment (andrei-kruhliakou-itechart) [inline Backend/CarRentalApp/CarRentalApp/Services/Identity/UserService.cs]: Why password matching with some external data is considered as a user validation?

File: Backend/CarRentalApp/CarRentalApp/appsettings.json [hunk 4]
```diff
@@ -5,5 +5,26 @@
       "Microsoft.AspNetCore": "Warning"
     }
   },
-  "AllowedHosts": "*"
-}
+  "AccessJwtConfig": {
+    "GenerationParameters": {
+      "Issuer": "loginServer",
+      "Audience": "clientsApp",
+      "LifeTimeSeconds": "300",
```
Comment (andrei-kruhliakou-itechart) [inline Backend/CarRentalApp/CarRentalApp/appsettings.json:12]: Why exactly this value?

Comment (Eugineer0) [reply Backend/CarRentalApp/CarRentalApp/appsettings.json:12]: Lifetime needs to be as short as possible to protect from token robbery. So if Bob steals Alice`s token, he will get only 5 min of access.

File: Backend/CarRentalApp/CarRentalApp/Services/Identity/UserService.cs (hunk 3 above)
Comment (Eugineer0) [reply Backend/CarRentalApp/CarRentalApp/Services/Identity/UserService.cs]: As for me. this method implements logic of "user checking". Now it delegates work only to password verification service, but in future, if "user checking" expands in something more than only password verification, I will have to change body of the method(its name and signature will stay untouchable).

Comment (andrei-kruhliakou-itechart) [reply Backend/CarRentalApp/CarRentalApp/Services/Identity/UserService.cs]: I would recommend to separate password validation from some validation (e.g. registration) and at least rename this method accordingly (e.g. ValidatePassword).

**Review by andrei-kruhliakou-itechart — APPROVED** [review]

**Review by andrei-kruhliakou-itechart — APPROVED** [review]

File: Backend/CarRentalApp/CarRentalApp/CarRentalApp.csproj [hunk 5]
```diff
@@ -7,11 +7,14 @@
...
-
-  <ItemGroup>
-    <Folder Include="Controllers\" />
-  </ItemGroup>
+    <PackageReference Include="BCrypt.Net-Core" Version="1.6.0" />
+    <PackageReference Include="Microsoft.AspNetCore.Authentication.JwtBearer" Version="6.0.1" />
+    <PackageReference Include="Microsoft.EntityFrameworkCore.SqlServer" Version="6.0.1" />
+    <PackageReference Include="Microsoft.EntityFrameworkCore.Tools" Version="6.0.1">
+      <PrivateAssets>all</PrivateAssets>
+      <IncludeAssets>runtime; build; native; contentfiles; analyzers; buildtransitive</IncludeAssets>
```
Comment (KseniyaMikhailiuk) [inline Backend/CarRentalApp/CarRentalApp/CarRentalApp.csproj:15]: what do these attributes mean?

File: Backend/CarRentalApp/CarRentalApp/Configuration/JWT/Access/AccessJwtConfig.cs [hunk 6]
```diff
@@ -0,0 +1,12 @@
+﻿using Microsoft.IdentityModel.Tokens;
+
+namespace CarRentalApp.Configuration.JWT.Access
+{
+    public class AccessJwtConfig
+    {
+        public const string Section = "AccessJwtConfig";
```
Comment (KseniyaMikhailiuk) [inline Backend/CarRentalApp/CarRentalApp/Configuration/JWT/Access/AccessJwtConfig.cs:7]: empty line?

File: Backend/CarRentalApp/CarRentalApp/Migrations/AuthenticationDbContextModelSnapshot.cs [hunk 7]
```diff
@@ -0,0 +1,60 @@
...
+                    b.Property<Guid>("Id")
+                        .ValueGeneratedOnAdd()
+                        .HasColumnType("uniqueidentifier");
+
+                    b.Property<DateTime>("Birthday")
+                        .HasColumnType("datetime2");
+
+                    b.Property<string>("Email")
+                        .IsRequired()
+                        .HasColumnType("nvarchar(max)");
```
Comment (KseniyaMikhailiuk) [inline Backend/CarRentalApp/CarRentalApp/Migrations/AuthenticationDbContextModelSnapshot.cs:36]: what is email maximum length?
are you sure that it is a good idea to have nvarchar(max) columns for email/username?

File: Backend/CarRentalApp/CarRentalApp/Migrations/AuthenticationDbContextModelSnapshot.cs [hunk 8]
```diff
@@ -0,0 +1,60 @@
...
+                    b.Property<string>("HashedPassword")
+                        .IsRequired()
+                        .HasColumnType("nvarchar(max)");
+
+                    b.Property<int>("Role")
+                        .HasColumnType("int");
+
+                    b.Property<string>("Salt")
+                        .IsRequired()
+                        .HasColumnType("nvarchar(max)");
```
Comment (KseniyaMikhailiuk) [inline Backend/CarRentalApp/CarRentalApp/Migrations/AuthenticationDbContextModelSnapshot.cs:47]: do you know the result length of salt and hash?

File: Backend/CarRentalApp/CarRentalApp/Services/Authentication/AuthenticationService.cs [hunk 9]
```diff
@@ -0,0 +1,44 @@
+﻿using System.IdentityModel.Tokens.Jwt;
+using CarRentalApp.Models.DTOs.Requests;
+using CarRentalApp.Models.DTOs.Responces;
+using CarRentalApp.Models.Entities;
+using CarRentalApp.Services.Identity;
+using CarRentalApp.Services.Token;
+
+namespace CarRentalApp.Services.Authentication
+{
+
```
Comment (KseniyaMikhailiuk) [inline Backend/CarRentalApp/CarRentalApp/Services/Authentication/AuthenticationService.cs:10]: useless line

File: Backend/CarRentalApp/CarRentalApp/Services/Identity/UserService.cs [hunk 10]
```diff
@@ -0,0 +1,33 @@
...
+        private UserRepository _userRepository;
+        private PasswordService _passwordService;
+
+        public UserService(UserRepository userRepository, PasswordService passwordService)
+        {
+            _userRepository = userRepository;
+            _passwordService = passwordService;
+        }
+
+        public async Task<User?> GetExistingUserAsync(UserLoginDTO user2Login)
```
Comment (KseniyaMikhailiuk) [inline Backend/CarRentalApp/CarRentalApp/Services/Identity/UserService.cs:19]: could you please use full variables name?

File: Backend/CarRentalApp/CarRentalApp/Services/Authentication/BCryptPasswordService.cs [hunk 11]
```diff
@@ -0,0 +1,15 @@
+﻿namespace CarRentalApp.Services.Authentication
+{
+    public class BCryptPasswordService : PasswordService
```
Comment (KseniyaMikhailiuk) [inline Backend/CarRentalApp/CarRentalApp/Services/Authentication/BCryptPasswordService.cs:3]: do you use it anywhere?

File: Backend/CarRentalApp/CarRentalApp/Contexts/AuthenticationDbContext.cs [hunk 12]
```diff
@@ -0,0 +1,15 @@
...
+using Microsoft.EntityFrameworkCore;
+
+namespace CarRentalApp.Models.Contexts
+{
+    public class AuthenticationDbContext : DbContext
+    {
+        public DbSet<User> Users { get; set; }        
+
+        public AuthenticationDbContext(DbContextOptions<AuthenticationDbContext> options)
+        : base(options)
```
Comment (KseniyaMikhailiuk) [inline Backend/CarRentalApp/CarRentalApp/Contexts/AuthenticationDbContext.cs:11]: indent

File: Backend/CarRentalApp/CarRentalApp/Contexts/AuthenticationDbContext.cs [hunk 13]
```diff
@@ -0,0 +1,15 @@
+﻿using CarRentalApp.Models.Entities;
+using Microsoft.EntityFrameworkCore;
+
+namespace CarRentalApp.Models.Contexts
+{
+    public class AuthenticationDbContext : DbContext
```
Comment (KseniyaMikhailiuk) [inline Backend/CarRentalApp/CarRentalApp/Contexts/AuthenticationDbContext.cs:6]: this context will be used only for authentication?

File: Backend/CarRentalApp/CarRentalApp/Services/Identity/UserService.cs [hunk 14]
```diff
@@ -0,0 +1,33 @@
...
+        {
+            return await _userRepository.GetByUsernameAsync(user2Login.Username);
```
Comment (KseniyaMikhailiuk) [inline Backend/CarRentalApp/CarRentalApp/Services/Identity/UserService.cs:21]: you can just return task without awaiting, as you don't need its result

File: Backend/CarRentalApp/CarRentalApp/Services/Authentication/ShaPasswordService .cs [hunk 15]
```diff
@@ -0,0 +1,30 @@
...
+    {
+        public override string DigestPassword(string password, string salt)
+        {
+            var hashFunc = SHA256.Create();
+
+            var data = Encoding.UTF8.GetBytes(password + salt);
+
+            var hashedPassword = hashFunc.ComputeHash(data);
+
+            return Convert.ToBase64String(hashedPassword);
```
Comment (KseniyaMikhailiuk) [inline Backend/CarRentalApp/CarRentalApp/Services/Authentication/ShaPasswordService .cs:16]: keep byte[]

**Pull request merged**